storage/innobase/dict/dict0dict.cc

```diff
@@ -1113,59 +1160,34 @@ void dict_table_t::add_to_cache()
 }
 
 /** Add a table definition to the data dictionary cache */
-inline void dict_sys_t::add(dict_table_t* table)
+inline void dict_sys_t::add(dict_table_t *table) noexcept
 {
-	ut_ad(!find(table));
-
-	ulint fold = my_crc32c(0, table->name.m_name,
-			       strlen(table->name.m_name));
-
-	table->autoinc_mutex.init();
-	table->lock_mutex_init();
-
-	/* Look for a table with the same name: error if such exists */
-	{
-		dict_table_t*	table2;
-		HASH_SEARCH(name_hash, &table_hash, fold,
-			    dict_table_t*, table2, ut_ad(table2->cached),
-			    !strcmp(table2->name.m_name, table->name.m_name));
-		ut_a(table2 == NULL);
-
-#ifdef UNIV_DEBUG
-		/* Look for the same table pointer with a different name */
-		HASH_SEARCH_ALL(name_hash, &table_hash,
-				dict_table_t*, table2, ut_ad(table2->cached),
-				table2 == table);
-		ut_ad(table2 == NULL);
-#endif /* UNIV_DEBUG */
-	}
-	HASH_INSERT(dict_table_t, name_hash, &table_hash, fold, table);
-
-	/* Look for a table with the same id: error if such exists */
-	hash_table_t* id_hash = table->is_temporary()
-		? &temp_id_hash : &table_id_hash;
-	const ulint id_fold = ut_fold_ull(table->id);
-	{
-		dict_table_t*	table2;
-		HASH_SEARCH(id_hash, id_hash, id_fold,
-			    dict_table_t*, table2, ut_ad(table2->cached),
-			    table2->id == table->id);
-		ut_a(table2 == NULL);
-
-#ifdef UNIV_DEBUG
-		/* Look for the same table pointer with a different id */
-		HASH_SEARCH_ALL(id_hash, id_hash,
-				dict_table_t*, table2, ut_ad(table2->cached),
-				table2 == table);
-		ut_ad(table2 == NULL);
-#endif /* UNIV_DEBUG */
-
-		HASH_INSERT(dict_table_t, id_hash, id_hash, id_fold, table);
-	}
+  ut_ad(!table->name_hash);
+  ut_ad(!table->id_hash);
+  table->autoinc_mutex.init();
+  table->lock_mutex_init();
+  dict_table_t **after;
+  after= reinterpret_cast<dict_table_t**>
+    (&table_hash.cell_get(my_crc32c(0, table->name.m_name,
+                                    strlen(table->name.m_name)))->node);
+  for (; *after; after = &(*after)->name_hash)
+  {
+    ut_ad((*after)->cached);
+    ut_a(strcmp((*after)->name.m_name, table->name.m_name));
+  }
+  *after= table;
 
-	UT_LIST_ADD_FIRST(table->can_be_evicted ? table_LRU : table_non_LRU,
-			  table);
-	ut_ad(dict_lru_validate());
+  after= reinterpret_cast<dict_table_t**>
+    (&(table->is_temporary() ? temp_id_hash : table_id_hash).
+     cell_get(ut_fold_ull(table->id))->node);
+  for (; *after; after = &(*after)->id_hash)
```

Comment:
Your comment is about f2fd2b81213edf7f64d1b074e2dcacb0d42ec8f4, which replaces `ut_fold_ull()` with an identity mapping on 64-bit systems. As far as I understand hashing, the main thing is that the slot in the hash table array is chosen by a modulus over the size of the hash table, which had better be a prime number. The `ut_fold_ull()` function was unnecessarily losing some entropy on 64-bit systems and potentially causing some premature loss of entropy, that is, causing some hash collision that could have been unavoidable. Any collision should be rather unlikely, because neither table or index identifiers are likely to be larger than `1U<<32`. The main win of refactoring `ut_fold_ull()` should be to reduce the size of the code and therefore to get a little performance improvement.

Related to this, there is also the function `ut_hash_ulint()`, which is performing an unnecessary-looking exclusive-or operation before calculating a modulus. Based on my reading of hash algorithms and of mysql/mysql-server@b11a175924194d574238f42068f09b15924ae2f8 we had better get rid of that function as well. While it does not seem to lose any entropy, it is an unnecessary obfuscation.